Add Test automation article for mobile

matijadurdek opened the pull request:
I'm adding a new article related to adding resource IDs to elements used in test automation on mobile. Initially, we wrote the article on one of our projects and decided to post it in the Dev process handbook.

@fgulan feel free to check it out and suggest any changes. Some notes:
- The title is currently `Test Automation (Mobile) - element IDs` - we can shorten it if needed. 
    - NOTE: We will add a section related to Web apps in the future, but the article is currently mobile-only.
- I added a short Table of ccontents at the beginning - we can remove it if it is not needed.

fgulan commented on General coding practices/Test Automation (Mobile) - element IDs.md:
```diff
@@ -0,0 +1,98 @@
+**Chapters:**
```
We can skip ToC on shorter chapters like this

matijadurdek replied:
Removed ✅ 
I added it initially so that mobile devs can easily find and jump to the paragraphs for their respective platforms - iOS or Android (XML vs Compose). So that they don't have to read through the whole text in order to find the section for their platform - because not all of it is relevant to them, just the sections for their respective platform.
It's a quick update, I can re-add it if you think it makes sense to have it 😄

fgulan commented on General coding practices/Test Automation (Mobile) - element IDs.md:
```diff
@@ -0,0 +1,98 @@
+
+### Accessibility Identifier
+
+For UI testing purposes Apple provides us with a so-called [accessibilityIdentifier](https://developer.apple.com/documentation/uikit/uiaccessibilityidentification/1623132-accessibilityidentifier). It is meant to "be used to uniquely identify a UI element in the scripts we write using the UI Automation interfaces". Accessibility identifiers aren't accessed by e.g. VoiceOver. They are part of the ` ...
```
```suggestion
For UI testing purposes Apple provides us with a so-called [accessibilityIdentifier](https://developer.apple.com/documentation/uikit/uiaccessibilityidentification/1623132-accessibilityidentifier). It is meant to "be used to uniquely identify a UI element in the scripts we write using the UI Automation interfaces". Accessibility identifiers aren't accessed by e.g. VoiceOver. They are part of the `UIAccessibilityIdentification` protocol, which consists of methods that associate a **unique identifier** with elements in a user interface.
```

No need to mention this, this is something that is default to a protocol definition

fgulan commented on General coding practices/Test Automation (Mobile) - element IDs.md:
```diff
@@ -0,0 +1,98 @@
+
+For UI testing purposes Apple provides us with a so-called [accessibilityIdentifier](https://developer.apple.com/documentation/uikit/uiaccessibilityidentification/1623132-accessibilityidentifier). It is meant to "be used to uniquely identify a UI element in the scripts we write using the UI Automation interfaces". Accessibility identifiers aren't accessed by e.g. VoiceOver. They are part of the ` ...
+
+`UIAccessibilityIdentification: var accessibilityIdentifier: String? { get set }`
```
This as well, you can omit it

matijadurdek replied:
Just to confirm - you meant the whole paragraph, not just the code example, right? I removed both the paragraph and code example, but left a small note regarding accessibility similar to the one that we have for Android
```
NOTE: accessibilityIdentifier isn't used for text-to-speech accessibility (accessed by e.g. VoiceOver).
```

fgulan replied:
Sorry, I wasn't specific - the paragraph was fine and necessary 😄 but the code example can be omitted. I've expected from GH to highlight only than line of code thingy but for some reason it highlighted the whole chapter. Sorry!

matijadurdek replied:
No worries, I thought the same because I expected GH to state which line is affected by the comment. 😄 
I reverted the removal and updated the text as per other comments. It should be okay now.

fgulan commented on General coding practices/Test Automation (Mobile) - element IDs.md:
```diff
@@ -0,0 +1,98 @@
+
+`UIAccessibilityIdentification: var accessibilityIdentifier: String? { get set }`
+
+The mobile developer needs to set accessibilityIdentifiers on elements that are being inspected as a part of test automation (e.g., buttons, input fields, list containers and items, titles, etc.). An accessibility identifier can be set using storyboards or programmatically. In our app, we should set it programmatically to have more readable code and to make changes easier if some identifier needs ...
```
```suggestion
The mobile developer needs to set `accessibilityIdentifier` on elements that are being inspected as a part of test automation (e.g., buttons, input fields, list containers and items, titles, etc.). An accessibility identifier can be set using storyboards or programmatically. In our app, we should set it programmatically to have more readable code and to make changes easier if some identifier needs to be updated.
```

fgulan replied:
```suggestion
The mobile developer needs to set accessibilityIdentifiers on elements that are being inspected as a part of test automation (e.g., buttons, input fields, list containers and items, titles, etc.). An accessibility identifier can be set using storyboards or programmatically.
```

I wouldn't insist on this one - each project has its own structure. If the project is exclusively and extensively using SB, developer will probably use SB for setting identifier as well which is from dev perspective totally fine.

fgulan commented on General coding practices/Test Automation (Mobile) - element IDs.md:
```diff
@@ -0,0 +1,98 @@
+
+The mobile developer needs to set accessibilityIdentifiers on elements that are being inspected as a part of test automation (e.g., buttons, input fields, list containers and items, titles, etc.). An accessibility identifier can be set using storyboards or programmatically. In our app, we should set it programmatically to have more readable code and to make changes easier if some identifier needs ...
+
+For list items that are not known in advance, the developer should implement an ID to the list container. After that, the testers should be able to get those elements as children of the container. Also, in case of adding IDs to individual list items that are not known in advance, it is alright if the same ID is implemented to them because the automation framework can locate such elements by combi ...
```
Maybe it's just me, but the `the developer` and `the tester` narrative throughout the chapter is a bit weird. It seems like we are doing some kind of a division between those two folks on the project. For example this one:

```
For list items that are not known in advance, the developer should implement an ID to the list container.
```

I would rework it a bit to

```
For list items that are not known in advance, developers should implement an ID into the list container.
```
 or
```
For list items that are not known in advance, ID should be added to te list container.
```

What do you think about it? I do not insist on this one!

matijadurdek replied:
Completely agree with you. 😄 
I updated it in the whole text and left "developer/tester" only in the cases where it is needed for semantic reasons. The rest is updated (either passive or generic forms are used). Feel free to check it out.

fgulan commented on General coding practices/Test Automation (Mobile) - element IDs.md:
```diff
@@ -0,0 +1,98 @@
+
+##### Unknown in advance
+
+For list items that are not known in advance, the developer should implement an ID to the list container. After that, the testers should be able to get those elements as children of the container. Also, in case of adding IDs to individual list items that are not known in advance, it is alright if the same ID is implemented to them because the automation framework can locate such elements by combi ...
```
```suggestion
For list items that are not known in advance, while developing ID should be added to the list container. After that, while testing we should be able to get those elements as children of the container. Also, in case of adding IDs to individual list items that are not known in advance, it is alright if the same ID is implemented to them because the automation framework can locate such elements by combining the ID and text of the element.
```

The same thing as above, what do you think?

matijadurdek replied:
Updated: https://github.com/infinum/devproc-handbook/pull/9#discussion_r1562873145

fgulan reviewed (commented):
I left you a few minor comments, but other than that really god job! Kudos! 🎉

matijadurdek commented:
> I left you a few minor comments, but other than that really god job! Kudos! 🎉

Thanks @fgulan for the detailed check and suggestions 💪 
I updated the article according to your suggestions. Feel free to check it out. 
Additionally, I updated some other things that I stumbled upon, and that didn't fit into the article as it was written a while ago on a project, so it had some sentences that weren't generic enough. 😄

fgulan commented:
@matijadurdek just a minor comment left on code example thingy, other than that - excellent chapter! 🎉 When you fix it, feel free to merge!

matijadurdek commented:
> @matijadurdek just a minor comment left on code example thingy, other than that - excellent chapter! 🎉 When you fix it, feel free to merge!

Merging it as agreed ✅
